core/store/src/trie/trie_storage.rs

```diff
@@ -426,6 +426,73 @@ impl TrieCachingStorage {
         key[8..].copy_from_slice(hash.as_ref());
         key
     }
+
+    /// Reads value if it is not in shard cache. Handles dropping the cache
+    /// lock. Either waits for prefetcher to fetch it or reads it from DB.
+    fn read_for_shard_cache_miss(
+        &self,
+        guard: std::sync::MutexGuard<TrieCacheInner>,
+        hash: &CryptoHash,
+    ) -> Result<Arc<[u8]>, StorageError> {
+        let val = if let Some(prefetcher) = &self.prefetch_api {
```

Comment:
nit: I would suggest inverting the logic and returning early to keep the interesting implementation less nested
```
let val = if let Some(prefetcher) = &self.prefetch_api else { 
  std::mem::drop(guard);
  self.read_from_db(hash)?
  return Ok(());
}
```